packages/kit/src/runtime/client/client.js

```diff
@@ -0,0 +1,1233 @@
+import { onMount, tick } from 'svelte';
+import { writable } from 'svelte/store';
+import { coalesce_to_error } from '../../utils/error.js';
+import { normalize } from '../load.js';
+import { normalize_path } from '../../utils/url';
+import {
+	create_updated_store,
+	find_anchor,
+	get_base_uri,
+	get_href,
+	initial_fetch,
+	notifiable_store,
+	scroll_state
+} from './utils';
+
+import Root from '__GENERATED__/root.svelte';
+import { routes, fallback } from '__GENERATED__/manifest.js';
+
+const SCROLL_KEY = 'sveltekit:scroll';
+const INDEX_KEY = 'sveltekit:index';
+
+// We track the scroll position associated with each history entry in sessionStorage,
+// rather than on history.state itself, because when navigation is driven by
+// popstate it's too late to update the scroll position associated with the
+// state we're navigating from
+
+/** @typedef {{ x: number, y: number }} ScrollPosition */
+/** @type {Record<number, ScrollPosition>} */
+let scroll_positions = {};
+try {
+	scroll_positions = JSON.parse(sessionStorage[SCROLL_KEY]);
+} catch {
+	// do nothing
+}
+
+/** @param {number} index */
+function update_scroll_positions(index) {
+	scroll_positions[index] = scroll_state();
+}
+
+/**
+ * @param {{
+ *   target: Element;
+ *   session: App.Session;
+ *   base: string;
+ *   trailing_slash: import('types').TrailingSlash;
+ * }} opts
+ * @returns {import('./types').Client}
+ */
+export function create_client({ target, session, base, trailing_slash }) {
+	/** @type {Map<string, import('./types').NavigationResult>} */
+	const cache = new Map();
+
+	/** @type {Set<string>} */
+	const invalid = new Set();
+
+	const stores = {
+		url: notifiable_store({}),
+		page: notifiable_store({}),
+		navigating: writable(/** @type {import('types').Navigation | null} */ (null)),
+		session: writable(session),
+		updated: create_updated_store()
+	};
+
+	/** @type {{id: string | null, promise: Promise<import('./types').NavigationResult | undefined> | null}} */
+	const loading = {
+		id: null,
+		promise: null
+	};
+
+	const callbacks = {
+		/** @type {Array<(opts: { from: URL, to: URL | null, cancel: () => void }) => void>} */
+		before_navigate: [],
+
+		/** @type {Array<(opts: { from: URL | null, to: URL }) => void>} */
+		after_navigate: []
+	};
+
+	/** @type {import('./types').NavigationState} */
+	let current = {
+		// @ts-ignore - we need the initial value to be null
```

Comment:
is this true? could we not just do `new URL(window.location.href)`?

Reply:
Conceptually, upon hydration we're navigating from `null` to `location.href`, and this is reflected in places like https://github.com/sveltejs/kit/blob/7373dd5396d6998749ac1766e359cfa28b523888/packages/kit/src/runtime/client/client.js#L396 and https://github.com/sveltejs/kit/blob/7373dd5396d6998749ac1766e359cfa28b523888/packages/kit/src/runtime/client/client.js#L564-L568